<a href="https://colab.research.google.com/github/PaulAlexInc/Cervical-Cancer-DataSet/blob/main/CervicalCancerDataSet.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


#Data Cleaning

Paul Alex<br>




[Cervical cancer (Risk Factors) Data Set](https://archive.ics.uci.edu/ml/datasets/Cervical+cancer+%28Risk+Factors%29#)

[Dataset](https://archive.ics.uci.edu/ml/machine-learning-databases/00383/)

In [ ]:
#importing libraries
import pandas as pd
import numpy as np

In [ ]:
data=pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/00383/risk_factors_cervical_cancer.csv")
data.head()

Age Number of sexual partners  ... Citology Biopsy
0   18                       4.0  ...        0      0
1   15                       1.0  ...        0      0
2   34                       1.0  ...        0      0
3   52                       5.0  ...        0      0
4   46                       3.0  ...        0      0

[5 rows x 36 columns]

###Check whether there are categorical variables(variables that can be divided into categories) in the above dataset

We need to understand that from the variable description and the values they take<br>

Generally when there are categorical variables, their presence is denoted by 1 and their absence is denoted by 0. If there are multiple categories, they're denoted as 1,2,3,etc

In [ ]:
#printing column names
data.columns

Index(['Age', 'Number of sexual partners', 'First sexual intercourse',
       'Num of pregnancies', 'Smokes', 'Smokes (years)', 'Smokes (packs/year)',
       'Hormonal Contraceptives', 'Hormonal Contraceptives (years)', 'IUD',
       'IUD (years)', 'STDs', 'STDs (number)', 'STDs:condylomatosis',
       'STDs:cervical condylomatosis', 'STDs:vaginal condylomatosis',
       'STDs:vulvo-perineal condylomatosis', 'STDs:syphilis',
       'STDs:pelvic inflammatory disease', 'STDs:genital herpes',
       'STDs:molluscum contagiosum', 'STDs:AIDS', 'STDs:HIV',
       'STDs:Hepatitis B', 'STDs:HPV', 'STDs: Number of diagnosis',
       'STDs: Time since first diagnosis', 'STDs: Time since last diagnosis',
       'Dx:Cancer', 'Dx:CIN', 'Dx:HPV', 'Dx', 'Hinselmann', 'Schiller',
       'Citology', 'Biopsy'],
      dtype='object')

In [ ]:
#information on the dataset
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 858 entries, 0 to 857
Data columns (total 36 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Age                                 858 non-null    int64 
 1   Number of sexual partners           858 non-null    object
 2   First sexual intercourse            858 non-null    object
 3   Num of pregnancies                  858 non-null    object
 4   Smokes                              858 non-null    object
 5   Smokes (years)                      858 non-null    object
 6   Smokes (packs/year)                 858 non-null    object
 7   Hormonal Contraceptives             858 non-null    object
 8   Hormonal Contraceptives (years)     858 non-null    object
 9   IUD                                 858 non-null    object
 10  IUD (years)                         858 non-null    object
 11  STDs                                858 non-null    object

In [ ]:
"""
<class 'pandas.core.frame.DataFrame'>
=> It is a pandas dataframe

RangeIndex: 858 entries, 0 to 857
=> here 0 to 857 is the index

dtypes: int64(10), object(26)
=> 10 columns of type int and 26 columns of type object

=> object means it is a string, list or array, etc.
usually it is a string
"""

"\n<class 'pandas.core.frame.DataFrame'>\n=> It is a pandas dataframe\n\nRangeIndex: 858 entries, 0 to 857\n=> here 0 to 857 is the index\n\ndtypes: int64(10), object(26)\n=> 10 columns of type int and 26 columns of type object\n\n=> object means it is a string, list or array, etc.\nusually it is a string\n"

here all 858 values are considered as non null since according to python, all missing values should be denoted by NaN(Not a number). Here its denoted by ?, so it is counting ? also as a value.

No explicit missing values present, they are present in the form of question marks

The datatypes are improper and need to be converted

There is a fillna function which can be used to fill the missing values, but before that we need to clean the data, convert the question marks(?) to NaN

We also need to convert the object datatypes to proper datatypes according to the variables<br>

In [ ]:
#In "isnull()" function False means not null
data.isnull()

Age  Number of sexual partners  ...  Citology  Biopsy
0    False                      False  ...     False   False
1    False                      False  ...     False   False
2    False                      False  ...     False   False
3    False                      False  ...     False   False
4    False                      False  ...     False   False
..     ...                        ...  ...       ...     ...
853  False                      False  ...     False   False
854  False                      False  ...     False   False
855  False                      False  ...     False   False
856  False                      False  ...     False   False
857  False                      False  ...     False   False

[858 rows x 36 columns]

In [ ]:
#Finding the no of missing values in each column => use .sum() function with isnull() function
data.isnull().sum()

Age                                   0
Number of sexual partners             0
First sexual intercourse              0
Num of pregnancies                    0
Smokes                                0
Smokes (years)                        0
Smokes (packs/year)                   0
Hormonal Contraceptives               0
Hormonal Contraceptives (years)       0
IUD                                   0
IUD (years)                           0
STDs                                  0
STDs (number)                         0
STDs:condylomatosis                   0
STDs:cervical condylomatosis          0
STDs:vaginal condylomatosis           0
STDs:vulvo-perineal condylomatosis    0
STDs:syphilis                         0
STDs:pelvic inflammatory disease      0
STDs:genital herpes                   0
STDs:molluscum contagiosum            0
STDs:AIDS                             0
STDs:HIV                              0
STDs:Hepatitis B                      0
STDs:HPV                              0


For filling up the null(missing) values, we need measures of central tendency<br>

For numerical variables<br>

if the variable is symmetric => mean, median and mode are equal
if it is skewed => mean is not a good measure, it becomes very large in case of large outliers and very small in case of small outliers, median is a better value, since it is not afftected.<br>

For categorical variables<br>

We can use only mode<br>

In [ ]:
# we need to convert the null values(question marks) in the dataset into compiler understandable null values(NaN)

In [ ]:
#accessing columns using column names
data.Biopsy#this method fails for column names with spaces

0      0
1      0
2      0
3      0
4      0
      ..
853    0
854    0
855    0
856    0
857    0
Name: Biopsy, Length: 858, dtype: int64

In [ ]:
#accessing columns using column names better method
data['Number of sexual partners']

0      4.0
1      1.0
2      1.0
3      5.0
4      3.0
      ... 
853    3.0
854    2.0
855    2.0
856    2.0
857    2.0
Name: Number of sexual partners, Length: 858, dtype: object

In [ ]:
data.Biopsy.unique()#unique function is used to get all the unique values in a column

array([0, 1])

In [ ]:
# unique values for all the columns
for col in data.columns:
  print(col)
  print(data[col].unique())

Age
[18 15 34 52 46 42 51 26 45 44 27 43 40 41 39 37 38 36 35 33 31 32 30 23
 28 29 20 25 21 24 22 48 19 17 16 14 59 79 84 47 13 70 50 49]
Number of sexual partners
['4.0' '1.0' '5.0' '3.0' '2.0' '6.0' '?' '7.0' '15.0' '8.0' '10.0' '28.0'
 '9.0']
First sexual intercourse
['15.0' '14.0' '?' '16.0' '21.0' '23.0' '17.0' '26.0' '20.0' '25.0' '18.0'
 '27.0' '19.0' '24.0' '32.0' '13.0' '29.0' '11.0' '12.0' '22.0' '28.0'
 '10.0']
Num of pregnancies
['1.0' '4.0' '2.0' '6.0' '3.0' '5.0' '?' '8.0' '7.0' '0.0' '11.0' '10.0']
Smokes
['0.0' '1.0' '?']
Smokes (years)
['0.0' '37.0' '34.0' '1.266972909' '3.0' '12.0' '?' '18.0' '7.0' '19.0'
 '21.0' '15.0' '13.0' '16.0' '8.0' '4.0' '10.0' '22.0' '14.0' '0.5' '11.0'
 '9.0' '2.0' '5.0' '6.0' '1.0' '32.0' '24.0' '28.0' '20.0' '0.16']
Smokes (packs/year)
['0.0' '37.0' '3.4' '2.8' '0.04' '0.5132021277' '2.4' '6.0' '?' '9.0'
 '1.6' '19.0' '21.0' '0.32' '2.6' '0.8' '15.0' '2.0' '5.7' '1.0' '3.3'
 '3.5' '12.0' '0.025' '2.75' '0.2' '1.4' '5.0' '2.1' '0.7' '1.2' 

In [ ]:
#converting ? to NaN
#Replace is a pandas dataframe function
#data.replace(element to replace,element to replace with)
data.replace('?', np.NaN, inplace=True)# in pandas dataframe, usually null values are given in th numpy library, "inplace=True" makes the change permanent

In [ ]:
data.head(100)

Age Number of sexual partners  ... Citology Biopsy
0    18                       4.0  ...        0      0
1    15                       1.0  ...        0      0
2    34                       1.0  ...        0      0
3    52                       5.0  ...        0      0
4    46                       3.0  ...        0      0
..  ...                       ...  ...      ...    ...
95   31                       4.0  ...        0      0
96   35                       5.0  ...        1      1
97   35                       1.0  ...        0      0
98   31                       1.0  ...        0      0
99   33                       3.0  ...        0      0

[100 rows x 36 columns]

Question marks have been replaced with compiler understandable null values(NaN)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 858 entries, 0 to 857
Data columns (total 36 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Age                                 858 non-null    int64 
 1   Number of sexual partners           832 non-null    object
 2   First sexual intercourse            851 non-null    object
 3   Num of pregnancies                  802 non-null    object
 4   Smokes                              845 non-null    object
 5   Smokes (years)                      845 non-null    object
 6   Smokes (packs/year)                 845 non-null    object
 7   Hormonal Contraceptives             750 non-null    object
 8   Hormonal Contraceptives (years)     750 non-null    object
 9   IUD                                 741 non-null    object
 10  IUD (years)                         741 non-null    object
 11  STDs                                753 non-null    object

In [ ]:
data.isnull().sum()

Age                                     0
Number of sexual partners              26
First sexual intercourse                7
Num of pregnancies                     56
Smokes                                 13
Smokes (years)                         13
Smokes (packs/year)                    13
Hormonal Contraceptives               108
Hormonal Contraceptives (years)       108
IUD                                   117
IUD (years)                           117
STDs                                  105
STDs (number)                         105
STDs:condylomatosis                   105
STDs:cervical condylomatosis          105
STDs:vaginal condylomatosis           105
STDs:vulvo-perineal condylomatosis    105
STDs:syphilis                         105
STDs:pelvic inflammatory disease      105
STDs:genital herpes                   105
STDs:molluscum contagiosum            105
STDs:AIDS                             105
STDs:HIV                              105
STDs:Hepatitis B                  

In [ ]:
787/858

0.9172494172494172

In [ ]:
#almost 92% of the data is missing in these 2 columns:
"""
STDs: Time since first diagnosis      787
STDs: Time since last diagnosis       787
"""

'\nSTDs: Time since first diagnosis      787\nSTDs: Time since last diagnosis       787\n'

Blank values dont always represent missing values

In [ ]:
"""
let us consider a
Train ticket dataset
Refund status: Refund given/Not given : 95% of the rows is blank

Do we have to remove the column Refund status?

But lets bring our attention to this point, when is refund generally given for train tickets?

When the ticket is cancelled
 
Isn't it possible that only 5% of the tickets were actually cancelled and 95% percent of the tickets were not cancelled?
If there is no cancellation and no refund to be given then the refund status can be blank

So now we have a question, do the blank values actually represent missing values?
It can also mean that the ticket were not cancelled

BLANK VALUES DO NOT ALWAYS REPRESENT MISSING VALUES 
so we need to be very cautious while deciding whether to drop a column
""" 


"""
But in our case it is explicitly mentioned that "Several patients decided not to answer some of the questions because of privacy concerns (missing values)."
so here, blank values do represent missing values
"""



'\nBut in our case it is explicitly mentioned that "Several patients decided not to answer some of the questions because of privacy concerns (missing values)."\nso here, blank values do represent missing values\n'

In [ ]:
#drop the columns
data.drop(['STDs: Time since first diagnosis','STDs: Time since last diagnosis'], axis=1, inplace=True)
data.shape

(858, 34)

In [ ]:
data.columns

Index(['Age', 'Number of sexual partners', 'First sexual intercourse',
       'Num of pregnancies', 'Smokes', 'Smokes (years)', 'Smokes (packs/year)',
       'Hormonal Contraceptives', 'Hormonal Contraceptives (years)', 'IUD',
       'IUD (years)', 'STDs', 'STDs (number)', 'STDs:condylomatosis',
       'STDs:cervical condylomatosis', 'STDs:vaginal condylomatosis',
       'STDs:vulvo-perineal condylomatosis', 'STDs:syphilis',
       'STDs:pelvic inflammatory disease', 'STDs:genital herpes',
       'STDs:molluscum contagiosum', 'STDs:AIDS', 'STDs:HIV',
       'STDs:Hepatitis B', 'STDs:HPV', 'STDs: Number of diagnosis',
       'Dx:Cancer', 'Dx:CIN', 'Dx:HPV', 'Dx', 'Hinselmann', 'Schiller',
       'Citology', 'Biopsy'],
      dtype='object')

In [ ]:
#list of numerical variables
num=['Age', 'Number of sexual partners', 'First sexual intercourse',
       'Num of pregnancies', 'Smokes (years)', 'Smokes (packs/year)', 'Hormonal Contraceptives (years)',
       'IUD (years)', 'STDs (number)', 'STDs: Number of diagnosis']

#list of categorical variables
cat=list(data.columns.difference(num))

In [ ]:
data.isnull().any(axis=1)
#No of rows having missing values in one or more of the columns
#False means no missing values

0      False
1      False
2       True
3      False
4      False
       ...  
853    False
854    False
855    False
856    False
857    False
Length: 858, dtype: bool

In [ ]:
data.isnull().any(axis=1).sum()


190

In [ ]:
190/858

0.22144522144522144

In [ ]:
#22% did not answer one or more answers

In [ ]:
"""
2 persons
a- no missing data for 2 columns
b- missing data for 2 columns

32 rows are same for two persons
still we can't impute the data for the missing rows like in 
other dataset, since this is a medical dataset(where anything is possible)

Generally in medical data it is not necessary that same senario should lean to same results, we try avoiding the imputation of missing data.

so we remove the 190 missing rows

"""

"\n2 persons\na- no missing data for 2 columns\nb- missing data for 2 columns\n\n32 rows are same for two persons\nstill we can't impute the data for the missing rows like in \nother dataset, since this is a medical dataset(where anything is possible)\n\nGenerally in medical data it is not necessary that same senario should lean to same results, we try avoiding the imputation of missing data.\n\nso we remove the 190 missing rows\n\n"

In [ ]:
data[data.isnull().any(axis=1) == True]#filter (select the columns in which rows have missing values in one or more of the columns)

Age Number of sexual partners  ... Citology Biopsy
2     34                       1.0  ...        0      0
9     44                       3.0  ...        0      0
18    42                       2.0  ...        0      0
19    40                       2.0  ...        0      0
37    37                       NaN  ...        0      0
..   ...                       ...  ...      ...    ...
819   14                       1.0  ...        0      0
820   24                       2.0  ...        0      0
833   34                       1.0  ...        0      0
836   37                       3.0  ...        0      0
837   31                       9.0  ...        0      0

[190 rows x 34 columns]

In [ ]:
rows_to_be_removed = list(data[data.isnull().any(axis=1) == True].index)#gets the index of the rows to be removed using the filter

In [ ]:
#copy of the datframe
data_temp=data.copy()
data_temp.drop(rows_to_be_removed, inplace=True)

In [ ]:
data_temp.isnull().any(axis=1).sum()

0

In [ ]:
data[cat].mode()


Biopsy  Citology  Dx  ...  STDs:vulvo-perineal condylomatosis  Schiller  Smokes
0       0         0   0  ...                                 0.0         0     0.0

[1 rows x 24 columns]

In [ ]:
data[cat].mode().iloc[0]
#iloc[0] is used to access the first row since we can't pass a dataframe. we need a series. Withou iloc[0], we'll get a dataframe

#.fillna fills for all the values

Biopsy                                  0
Citology                                0
Dx                                      0
Dx:CIN                                  0
Dx:Cancer                               0
Dx:HPV                                  0
Hinselmann                              0
Hormonal Contraceptives               1.0
IUD                                   0.0
STDs                                  0.0
STDs:AIDS                             0.0
STDs:HIV                              0.0
STDs:HPV                              0.0
STDs:Hepatitis B                      0.0
STDs:cervical condylomatosis          0.0
STDs:condylomatosis                   0.0
STDs:genital herpes                   0.0
STDs:molluscum contagiosum            0.0
STDs:pelvic inflammatory disease      0.0
STDs:syphilis                         0.0
STDs:vaginal condylomatosis           0.0
STDs:vulvo-perineal condylomatosis    0.0
Schiller                                0
Smokes                            

In [ ]:
#imputation for categorical variables

data[cat].mode()

Biopsy  Citology  Dx  ...  STDs:vulvo-perineal condylomatosis  Schiller  Smokes
0       0         0   0  ...                                 0.0         0     0.0

[1 rows x 24 columns]

In [ ]:
data[cat].mode().values

array([[0, 0, 0, 0, 0, 0, 0, '1.0', '0.0', '0.0', '0.0', '0.0', '0.0',
        '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', '0.0', 0,
        '0.0']], dtype=object)

In [ ]:
data[cat]=data[cat].astype('float')
data[cat]

Biopsy  Citology  ...  Schiller  Smokes
0       0.0       0.0  ...       0.0     0.0
1       0.0       0.0  ...       0.0     0.0
2       0.0       0.0  ...       0.0     0.0
3       0.0       0.0  ...       0.0     1.0
4       0.0       0.0  ...       0.0     0.0
..      ...       ...  ...       ...     ...
853     0.0       0.0  ...       0.0     0.0
854     0.0       0.0  ...       0.0     0.0
855     0.0       1.0  ...       0.0     0.0
856     0.0       0.0  ...       0.0     0.0
857     0.0       0.0  ...       0.0     0.0

[858 rows x 24 columns]

In [ ]:
data.fillna(data[cat].mode().iloc[0], inplace=True)#we are using the row value to impute

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 858 entries, 0 to 857
Data columns (total 34 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Age                                 858 non-null    int64  
 1   Number of sexual partners           832 non-null    object 
 2   First sexual intercourse            851 non-null    object 
 3   Num of pregnancies                  802 non-null    object 
 4   Smokes                              858 non-null    float64
 5   Smokes (years)                      845 non-null    object 
 6   Smokes (packs/year)                 845 non-null    object 
 7   Hormonal Contraceptives             858 non-null    float64
 8   Hormonal Contraceptives (years)     750 non-null    object 
 9   IUD                                 858 non-null    float64
 10  IUD (years)                         741 non-null    object 
 11  STDs                                858 non-n

In [ ]:
data[cat].fillna(data[cat].mode().iloc[0]).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 858 entries, 0 to 857
Data columns (total 24 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Biopsy                              858 non-null    float64
 1   Citology                            858 non-null    float64
 2   Dx                                  858 non-null    float64
 3   Dx:CIN                              858 non-null    float64
 4   Dx:Cancer                           858 non-null    float64
 5   Dx:HPV                              858 non-null    float64
 6   Hinselmann                          858 non-null    float64
 7   Hormonal Contraceptives             858 non-null    float64
 8   IUD                                 858 non-null    float64
 9   STDs                                858 non-null    float64
 10  STDs:AIDS                           858 non-null    float64
 11  STDs:HIV                            858 non-n

We see that in all the categorical variables we have no missing values

Lets change the datatype now using astype() function

In [ ]:
#The changes will be reflected in data_temp not in data

In [ ]:
data[cat]=data[cat].astype('bool')

In [ ]:
#checking unique
for col in cat:
  print(col)
  print(data[col].unique())

Biopsy
[False  True]
Citology
[False  True]
Dx
[False  True]
Dx:CIN
[False  True]
Dx:Cancer
[False  True]
Dx:HPV
[False  True]
Hinselmann
[False  True]
Hormonal Contraceptives
[False  True]
IUD
[False  True]
STDs
[False  True]
STDs:AIDS
[False]
STDs:HIV
[False  True]
STDs:HPV
[False  True]
STDs:Hepatitis B
[False  True]
STDs:cervical condylomatosis
[False]
STDs:condylomatosis
[False  True]
STDs:genital herpes
[False  True]
STDs:molluscum contagiosum
[False  True]
STDs:pelvic inflammatory disease
[False  True]
STDs:syphilis
[False  True]
STDs:vaginal condylomatosis
[False  True]
STDs:vulvo-perineal condylomatosis
[False  True]
Schiller
[False  True]
Smokes
[False  True]


In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 858 entries, 0 to 857
Data columns (total 34 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Age                                 858 non-null    int64 
 1   Number of sexual partners           832 non-null    object
 2   First sexual intercourse            851 non-null    object
 3   Num of pregnancies                  802 non-null    object
 4   Smokes                              858 non-null    bool  
 5   Smokes (years)                      845 non-null    object
 6   Smokes (packs/year)                 845 non-null    object
 7   Hormonal Contraceptives             858 non-null    bool  
 8   Hormonal Contraceptives (years)     750 non-null    object
 9   IUD                                 858 non-null    bool  
 10  IUD (years)                         741 non-null    object
 11  STDs                                858 non-null    bool  

In [ ]:
"""
To impute the numerical values, we need to know the shape of the distribution to choose which measure of central
tendency to use.

Is there a measure which we can use to identify whether the distribution is skewed or symmetric, without using the shape of it
or without visualising it?

Yes, median

"""

'\nTo impute the numerical values, we need to know the shape of the distribution to choose which measure of central\ntendency to use.\n\nIs there a measure which we can use to identify whether the distribution is skewed or symmetric, without using the shape of it\nor without visualising it?\n\nYes, median\n\n'

In [ ]:
#for symmetric--> mean
#for skewed -->median

In [ ]:
#how do we know that?

#using describe() function

In [ ]:
#statistical summary of the dataset
data[num].describe()

Age  STDs: Number of diagnosis
count  858.000000                 858.000000
mean    26.820513                   0.087413
std      8.497948                   0.302545
min     13.000000                   0.000000
25%     20.000000                   0.000000
50%     25.000000                   0.000000
75%     32.000000                   0.000000
max     84.000000                   3.000000

In [ ]:
#why are we getting only two columns when there are actually 10 values

In [ ]:
len(num)

10

In [ ]:
#check the data types above

In [ ]:
num

['Age',
 'Number of sexual partners',
 'First sexual intercourse',
 'Num of pregnancies',
 'Smokes (years)',
 'Smokes (packs/year)',
 'Hormonal Contraceptives (years)',
 'IUD (years)',
 'STDs (number)',
 'STDs: Number of diagnosis']

In [ ]:
#many numerical variables are still in the form of objects. Operations like mean,median,etc work on int type only
#so before describe we need to convert their datatype

In [ ]:
# changing the datatype to float
data[num]=data[num].astype('float')

In [ ]:
data[num].describe()

Age  ...  STDs: Number of diagnosis
count  858.000000  ...                 858.000000
mean    26.820513  ...                   0.087413
std      8.497948  ...                   0.302545
min     13.000000  ...                   0.000000
25%     20.000000  ...                   0.000000
50%     25.000000  ...                   0.000000
75%     32.000000  ...                   0.000000
max     84.000000  ...                   3.000000

[8 rows x 10 columns]

In [ ]:
#imputation
#we can use only mode for categorical variables

In [ ]:
"""
What information does 25% represent?

25,20 and 75 represent the quartiles

=> 25% represents the first quartile

What information does 50% represent?
=> 50% represents the median or the second quartile

What information does 75% represent?
=> 25% represents the third quartile
"""

'\nWhat information does 25% represent?\n\n25,20 and 75 represent the quartiles\n\n=> 25% represents the first quartile\n\nWhat information does 50% represent?\n=> 50% represents the median or the second quartile\n\nWhat information does 75% represent?\n=> 25% represents the third quartile\n'

In [ ]:
#comparing the mean and median, if they are approximately same, the distribution can be considered to be symmetric

#if they are very far apart, the distribution can be considered to be skewed

#The differece between the mean and median can be used to determine the degree of skewness, if difference is very large, more skewed

#In datascience the scenario mean==median is very rare

In [ ]:
"""
in a range 1-10, is a diff of 2 significant?i.e 4 and 6 are significantly different?  yes (skewed)
in a range 1-10000, is a diff of 2 significant? i.e 10000 and 10002? no, we can consider it to be symmetric
"""

'\nin a range 1-10, is a diff of 2 significant?i.e 4 and 6 are significantly different?  yes (skewed)\nin a range 1-10000, is a diff of 2 significant? i.e 10000 and 10002? no, we can consider it to be symmetric\n'

Checking each variable whether its symmetric or skewed<br>
- Age<br>
    mean	26.820513<br>
    median	25.000000<br>
    difference = mean-median=>1.82<br>
    range - max-min => 84-13 => 71<br>
-- It is approximately symmetric(very slightly skewed)
<br>
<br>
- Number of sexual partners<br>
    mean	2.527644	<br>
    median	2.000000	<br>
    difference = mean-median=>2.52<br>
    range - max-min => 27<br>
-- It is approximately symmetric(very slightly skewed)
<br>
<br>
- First sexual intercourse<br>
    mean	16.995300	<br>
    median	17.000000<br>
    difference = mean-median=>0.01<br>
    range - max-min => 22<br>

-- It is approximately symmetric(very slightly skewed)
<br>
<br>
- Num of pregnancies	<br>
    mean	2.275561<br>
    median	2.000000<br>
    difference = mean-median=>0.27<br>
    range - max-min => 11<br>
-- It is approximately symmetric(very slightly skewed)

- Smokes(years): Skewed<br>
- Smokes(packs/year): Skewed<br>
- Hormonal Contraceptives (years): Skewed<br>
- IUD (years): Skewed<br>
- STDs (number): Skewed<br>
- STDs: Number of diagnosis: Skewed




In [ ]:
# imputation of the numerical variables with median
data.fillna(data[num].median().iloc[0], inplace=True)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 858 entries, 0 to 857
Data columns (total 34 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Age                                 858 non-null    float64
 1   Number of sexual partners           858 non-null    float64
 2   First sexual intercourse            858 non-null    float64
 3   Num of pregnancies                  858 non-null    float64
 4   Smokes                              858 non-null    bool   
 5   Smokes (years)                      858 non-null    float64
 6   Smokes (packs/year)                 858 non-null    float64
 7   Hormonal Contraceptives             858 non-null    bool   
 8   Hormonal Contraceptives (years)     858 non-null    float64
 9   IUD                                 858 non-null    bool   
 10  IUD (years)                         858 non-null    float64
 11  STDs                                858 non-n

In [ ]:
data[num]


Age  Number of sexual partners  ...  STDs (number)  STDs: Number of diagnosis
0    18.0                        4.0  ...            0.0                        0.0
1    15.0                        1.0  ...            0.0                        0.0
2    34.0                        1.0  ...            0.0                        0.0
3    52.0                        5.0  ...            0.0                        0.0
4    46.0                        3.0  ...            0.0                        0.0
..    ...                        ...  ...            ...                        ...
853  34.0                        3.0  ...            0.0                        0.0
854  32.0                        2.0  ...            0.0                        0.0
855  25.0                        2.0  ...            0.0                        0.0
856  33.0                        2.0  ...            0.0                        0.0
857  29.0                        2.0  ...            0.0                        0.0

[858 rows x 10 columns]

In [ ]:
# unique values for all the columns
for col in data.columns:
  print(col)
  print(data[col].unique())

Age
[18. 15. 34. 52. 46. 42. 51. 26. 45. 44. 27. 43. 40. 41. 39. 37. 38. 36.
 35. 33. 31. 32. 30. 23. 28. 29. 20. 25. 21. 24. 22. 48. 19. 17. 16. 14.
 59. 79. 84. 47. 13. 70. 50. 49.]
Number of sexual partners
[ 4.  1.  5.  3.  2.  6. 25.  7. 15.  8. 10. 28.  9.]
First sexual intercourse
[15. 14. 25. 16. 21. 23. 17. 26. 20. 18. 27. 19. 24. 32. 13. 29. 11. 12.
 22. 28. 10.]
Num of pregnancies
[ 1.  4.  2.  6.  3.  5. 25.  8.  7.  0. 11. 10.]
Smokes
[False  True]
Smokes (years)
[ 0.         37.         34.          1.26697291  3.         12.
 25.         18.          7.         19.         21.         15.
 13.         16.          8.          4.         10.         22.
 14.          0.5        11.          9.          2.          5.
  6.          1.         32.         24.         28.         20.
  0.16      ]
Smokes (packs/year)
[0.00000000e+00 3.70000000e+01 3.40000000e+00 2.80000000e+00
 4.00000000e-02 5.13202128e-01 2.40000000e+00 6.00000000e+00
 2.50000000e+01 9.00000000e+00 1.60000

We have a problem here, for some variables like STDs for instance

before replacing '?'<br>
STDs<br>
['0.0' '1.0' '?']<br>

0 should be false 1 should be true<br>

but what we are getting is<br>
STDs<br>
[ True]<br>

This is because we did the imputation before changing the datatype<br>

so before the imputation we're going to change thw datatype of the categorical columns to int

In [ ]:
# in compiler 0 is considered as false, EVERYTHING ELSE is considered true. Here while imputing we need to convert everfything as int, because even if 0 is stored as string we'll get true

In [ ]:
#cervical and aids

In [ ]:
# dropping the 2 columns which have only one value for the entire dataset
data.drop(['STDs:cervical condylomatosis', 'STDs:AIDS'], axis=1, inplace=True)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 858 entries, 0 to 857
Data columns (total 32 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Age                                 858 non-null    float64
 1   Number of sexual partners           858 non-null    float64
 2   First sexual intercourse            858 non-null    float64
 3   Num of pregnancies                  858 non-null    float64
 4   Smokes                              858 non-null    bool   
 5   Smokes (years)                      858 non-null    float64
 6   Smokes (packs/year)                 858 non-null    float64
 7   Hormonal Contraceptives             858 non-null    bool   
 8   Hormonal Contraceptives (years)     858 non-null    float64
 9   IUD                                 858 non-null    bool   
 10  IUD (years)                         858 non-null    float64
 11  STDs                                858 non-n

In [ ]:
data.to_csv("CervicalCancer_Cleaned", index=False)

#Data Encoding

if we have :<br>
2 categories: Presence -> 1, Absence -> 0<br>
5 categories: 1, 2, 3, 4, 5 -> 1 to n or 0 to n-1 -> Label Encoding: We can only use it for target variables(Variables that need to be predicted). We cannot use it on predictor variables <br>

Because when we write them as numbers like 1,2,3,4,5 ,there is an order in them. 5 is greater than 1. When we multiply a value by 5 or add a value 5 it is obviously going to have a greater effect on the result rather than the value 1<br>

Suppose we take 2<br>
- we multiply 1 to it
- we multiply 5 to it

When do we get a greater result?<br>
The second case (2 x 5=10)<br>

If there is an ordering in this way, some categories get more preference than the others, which we do not want to happen<br>

So what we do is<br>
For 5 categories we create 5 more columns and for each column we consider values as 0 and 1<br>
Category 1: Presence -> 1, Absence -> 0<br>
Category 2: Presence -> 1, Absence -> 0<br>
Category 3: Presence -> 1, Absence -> 0<br>
Category 4: Presence -> 1, Absence -> 0<br>
Category 5: Presence -> 1, Absence -> 0<br>

This process is called -> One Hot Encoding<br>

In [ ]:
cat_data = data['STDs: Number of diagnosis']
cat_data

0      0.0
1      0.0
2      0.0
3      0.0
4      0.0
      ... 
853    0.0
854    0.0
855    0.0
856    0.0
857    0.0
Name: STDs: Number of diagnosis, Length: 858, dtype: float64

In [ ]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder# fit_transfer after this
encoder=OneHotEncoder()

In [ ]:
encoder.fit(np.array(cat_data).reshape(-1, 1))

OneHotEncoder(categories='auto', drop=None, dtype=<class 'numpy.float64'>,
              handle_unknown='error', sparse=True)

In [ ]:
result = encoder.transform(np.array(cat_data).reshape(-1, 1))

In [ ]:
encoder.fit_transform(np.array(cat_data).reshape(-1, 1))#In one line

<858x4 sparse matrix of type '<class 'numpy.float64'>'
	with 858 stored elements in Compressed Sparse Row format>

In [ ]:
print(result)

  (0, 0)	1.0
  (1, 0)	1.0
  (2, 0)	1.0
  (3, 0)	1.0
  (4, 0)	1.0
  (5, 0)	1.0
  (6, 0)	1.0
  (7, 0)	1.0
  (8, 0)	1.0
  (9, 0)	1.0
  (10, 0)	1.0
  (11, 0)	1.0
  (12, 0)	1.0
  (13, 0)	1.0
  (14, 0)	1.0
  (15, 0)	1.0
  (16, 0)	1.0
  (17, 0)	1.0
  (18, 0)	1.0
  (19, 0)	1.0
  (20, 0)	1.0
  (21, 1)	1.0
  (22, 1)	1.0
  (23, 0)	1.0
  (24, 0)	1.0
  :	:
  (833, 0)	1.0
  (834, 0)	1.0
  (835, 0)	1.0
  (836, 0)	1.0
  (837, 0)	1.0
  (838, 0)	1.0
  (839, 0)	1.0
  (840, 0)	1.0
  (841, 0)	1.0
  (842, 0)	1.0
  (843, 0)	1.0
  (844, 0)	1.0
  (845, 0)	1.0
  (846, 0)	1.0
  (847, 0)	1.0
  (848, 0)	1.0
  (849, 0)	1.0
  (850, 0)	1.0
  (851, 0)	1.0
  (852, 0)	1.0
  (853, 0)	1.0
  (854, 0)	1.0
  (855, 0)	1.0
  (856, 0)	1.0
  (857, 0)	1.0


In [ ]:
pd.get_dummies(data, columns=['STDs: Number of diagnosis'])#EASIER METHOD OF IMPLEMENTING ONEHOTENCODING, SHORTER CODE

Age  ...  STDs: Number of diagnosis_3.0
0    18.0  ...                              0
1    15.0  ...                              0
2    34.0  ...                              0
3    52.0  ...                              0
4    46.0  ...                              0
..    ...  ...                            ...
853  34.0  ...                              0
854  32.0  ...                              0
855  25.0  ...                              0
856  33.0  ...                              0
857  29.0  ...                              0

[858 rows x 35 columns]

#Data Scaling

Problem: Predict the price of house based on its features

We have two features<br>
Area of the house(in sq ft), No of bedrooms<br>
- Area: 10000, 20000, 15000, etc<br>
- No of bedrooms: 1,2,3,etc.<br>

Algorithms are mainly mathematical, if we multiply a number by 10000 and if we multiply a number by 1, obviously 10000 will have a larger influence<br>

- 2 x 10000=20000
- 2 x 1=2

So the model will give more preference to the area variable just because it has higher value<br>

The model might think that the area variable is more important than the number of bedrooms <br>

So we change the value of the variables to the same range<br>


In [ ]:
data[num].describe()

Age  ...  STDs: Number of diagnosis
count  858.000000  ...                 858.000000
mean    26.820513  ...                   0.087413
std      8.497948  ...                   0.302545
min     13.000000  ...                   0.000000
25%     20.000000  ...                   0.000000
50%     25.000000  ...                   0.000000
75%     32.000000  ...                   0.000000
max     84.000000  ...                   3.000000

[8 rows x 10 columns]

#Standard Scalar

(X-Mean(X))/Std Dev(X)

formula is similar to that of Z distribution

- widely used

In [ ]:
from sklearn.preprocessing import StandardScaler
scalar = StandardScaler()

In [ ]:
std_scaled_data = scalar.fit_transform(data[num])

In [ ]:
pd.DataFrame(std_scaled_data, columns=num).describe()

Age  ...  STDs: Number of diagnosis
count  8.580000e+02  ...               8.580000e+02
mean   3.524117e-16  ...              -9.264798e-17
std    1.000583e+00  ...               1.000583e+00
min   -1.627284e+00  ...              -2.890930e-01
25%   -8.030752e-01  ...              -2.890930e-01
50%   -2.143546e-01  ...              -2.890930e-01
75%    6.098541e-01  ...              -2.890930e-01
max    6.732547e+00  ...               9.632579e+00

[8 rows x 10 columns]

In [ ]:
#After running standard scalar
#Mean is approx 0
#Std is approx 1

#Min Max Scalar

[Max(X)-X]/[Max(X)-Min(X)]

X is Max => 0<br>
X is Min => 1<br>


- Used if there are variables which shouldn't have negative values
- If we want a probability(always lying between 0 and 1) or a proper fraction

In [ ]:
from sklearn.preprocessing import MinMaxScaler
scaler=MinMaxScaler()

In [ ]:
min_max_scaled_data = scaler.fit_transform(data[num])

In [ ]:
pd.DataFrame(min_max_scaled_data, columns=num).describe()

Age  ...  STDs: Number of diagnosis
count  858.000000  ...                 858.000000
mean     0.194655  ...                   0.029138
std      0.119689  ...                   0.100848
min      0.000000  ...                   0.000000
25%      0.098592  ...                   0.000000
50%      0.169014  ...                   0.000000
75%      0.267606  ...                   0.000000
max      1.000000  ...                   1.000000

[8 rows x 10 columns]

In [ ]:
#After predicting and before displaying it to the user, we need to descale it, there we can't use fit_transform, we need to first fit then transform

#Because then we will have to use the fitted scalar to inverse transorm the output values

In [ ]:
for uni in data.columns:
  print(uni)
  print(data[uni].unique())

Age
[18. 15. 34. 52. 46. 42. 51. 26. 45. 44. 27. 43. 40. 41. 39. 37. 38. 36.
 35. 33. 31. 32. 30. 23. 28. 29. 20. 25. 21. 24. 22. 48. 19. 17. 16. 14.
 59. 79. 84. 47. 13. 70. 50. 49.]
Number of sexual partners
[ 4.  1.  5.  3.  2.  6. 25.  7. 15.  8. 10. 28.  9.]
First sexual intercourse
[15. 14. 25. 16. 21. 23. 17. 26. 20. 18. 27. 19. 24. 32. 13. 29. 11. 12.
 22. 28. 10.]
Num of pregnancies
[ 1.  4.  2.  6.  3.  5. 25.  8.  7.  0. 11. 10.]
Smokes
[False  True]
Smokes (years)
[ 0.         37.         34.          1.26697291  3.         12.
 25.         18.          7.         19.         21.         15.
 13.         16.          8.          4.         10.         22.
 14.          0.5        11.          9.          2.          5.
  6.          1.         32.         24.         28.         20.
  0.16      ]
Smokes (packs/year)
[0.00000000e+00 3.70000000e+01 3.40000000e+00 2.80000000e+00
 4.00000000e-02 5.13202128e-01 2.40000000e+00 6.00000000e+00
 2.50000000e+01 9.00000000e+00 1.60000